File: motor.m
classdef motor < handle 
    properties 
        % Motor Spec's
        Ra = 4;                         % Resistance of motor, Ohms
        La = 0.1;                       % Inductance of motor, H
        Kt = 0.35;                      % torque constant, Nm/A
        Ke = 0.35;                      % EMF constant, V/rad/s
        bs = 0.008;                     % Viscous torque, Nm 
        Jm = 0.005;                     % Moment of Inertia for motor, kg m^2
        fric_constant;% Friction acting on the wheel, Nm
        
        % Dynamic Variables
        I; % Amps
        w; % rad/s
        t; % Nm
        
        % Desired Velocity 
        Va
        
    end
    
    methods
        
        function obj = motor(I, w, t, Va, xtra_fric)
            obj.I = I;
            obj.w = w;
            obj.t = t;
            obj.Va = Va;
            obj.fric_constant = 0.002 + xtra_fric;
        end
        
        function [Idot, wdot, tdot] = update(obj, motorInputs, Va, stepsize, xtra_fric)
            obj.fric_constant = 0.002+xtra_fric; % UpdateFriction
            obj.I = motorInputs(1);
            obj.w = motorInputs(2);
            obj.t = motorInputs(3);
         
            Idot = ((-obj.Ra*obj.I)-(obj.Ke*obj.w)+ Va)/obj.La;              % di/dt 
            tau_friction = obj.fric_constant*obj.w;
            wdot = ((obj.Kt*obj.I)-(obj.bs*obj.w)-tau_friction)/obj.Jm;      % Angular Acceleration

            eff = (abs(obj.I)*-0.1330)+0.6;
            if (Va == 0)
                tdot = (0-obj.t)/stepsize;
            else
                tdot = (obj.Kt*obj.I*eff-obj.t)/stepsize;                          % Torque
            end

        end
            
    end
end

File: src/DynamicModels/motor.cpp
#include "motor.h"
#include <cmath>

motor::motor(const float current, const float ang_velo, const float torque, const float frict) :
I(current), w(ang_velo), t(torque), fric_constant(frict + 0.002)
{

}

Vector3f motor::update(const float &v_desired, const float &stepsize)
{
    const float tau_friction = fric_constant*w;

    Vector3f Xdot_motor;
    const float Idot = ((-Ra*I)-(Ke*w)+ Va)/La;              //  di/dt 
    const float wdot = ((Kt*I)-(bs*w)-tau_friction)/Jm;      // Angular Acceleration

    const eff = (std::abs(I)*-0.1330)+0.6; // Unsure where the "magic" numbers come from

    const float tdot = ((v_desired == 0) ? (-t)/stepsize : (Kt*I*eff-t)/stepsize); // dT/dt

    Xdot_motor << Idot, wdot, tdot;

    return Xdot_motor;
}

Vector3f motor::get_states()
{
    Vector3f states;

    states << I, w, t;
    
    return states;
}

void motor::set_states(Vector3f &states)
{
    I = states(0);
    w = states(1);
    t = states(2);
}